src/core/common/pool.hpp

```diff
@@ -0,0 +1,185 @@
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for a generic object pool.
+ */
+
+#ifndef POOL_HPP_
+#define POOL_HPP_
+
+#include "openthread-core-config.h"
+
+#include "common/linked_list.hpp"
+#include "common/non_copyable.hpp"
+
+namespace ot {
+
+class Instance;
+
+/**
+ * @addtogroup core-pool
+ *
+ * @brief
+ *   This module includes definitions for OpenThread object pool.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This template class represents an object pool.
+ *
+ * The template type `Type` should provide `GetNext() and `SetNext()` so that it can be added to a linked list.
+ *
+ * The template parameter `kPoolSize` specifies maximum number of objects in the pool.
+ *
+ */
+template <class Type, uint16_t kPoolSize> class Pool : private NonCopyable
+{
+public:
+    /**
+     * This constructor initializes the pool.
+     *
+     */
+    Pool(void)
+        : mFreeList()
+    {
+        for (Type &entry : mPool)
+        {
+            mFreeList.Push(entry);
+        }
+    }
+
+    /**
+     * This constructor initializes the pool.
+     *
+     * This constructor version requires the `Type` class to provide method `void Init(Instance &)` to initialize
+     * each `Type` entry object. This can be realized by the `Type` class inheriting from `InstaceLocatorInit()`.
+     *
+     * @param[in] aInstance   A reference to the OpenThread instance.
+     *
+     */
+    Pool(Instance &aInstance)
+        : mFreeList()
+    {
+        for (Type &entry : mPool)
+        {
+            entry.Init(aInstance);
+            mFreeList.Push(entry);
+        }
+    }
+
+    /**
+     * This method allocates a new object from the pool.
+     *
+     * @returns A pointer to the newly allocated object, or nullptr if all entries from the pool are already allocated.
+     *
+     */
+    Type *Allocate(void) { return mFreeList.Pop(); }
+
+    /**
+     * This method frees a previously allocated object.
+     *
+     * The @p aEntry MUST be an entry from the pool previously allocated using `Allocate()` method and not yet freed.
+     * An already freed entry MUST not be freed again.
+     *
+     * @param[in]  aEntry   The pool object entry to free.
+     *
+     */
+    void Free(Type &aEntry) { mFreeList.Push(aEntry); }
+
+    /**
+     * This method returns the pool size.
+     *
+     * @returns The pool size (maximum number of objects in the pool).
+     *
+     */
+    uint16_t GetSize(void) const { return kPoolSize; }
+
+    /**
+     * This method indicates whether or not a given `Type` object is from the pool.
+     *
+     * @param[in]  aObject   A reference to a `Type` object.
+     *
+     * @retval TRUE if @p aObject is from the pool.
+     * @retval FALSE if @p aObject is not from the pool.
+     *
+     */
+    bool Contains(const Type &aObject) const { return (&mPool[0] <= &aObject) && (&aObject < OT_ARRAY_END(mPool)); }
```

Comment:
```suggestion
    bool IsPoolEntry(const Type &aObject) const { return (&mPool[0] <= &aObject) && (&aObject < OT_ARRAY_END(mPool)); }
```
It's reasonable to think that only unallocated entries are `Contained` in the pool.
Better to rename it?

Reply:
Good suggestion. Thanks. Changed it to `IsPoolEntry()` in the new pushed commit.